acnportal/acnsim/network/charging_network.py

```diff
@@ -163,20 +186,88 @@ def update_pilots(self, pilots, i, period):
+            t (List[int]): List of time indices for which to calculate aggregate current. If None, 
+                calculates aggregate currents for all timesteps.
+            linear (bool): If True, linearize all constraints to a more conservative but easier to compute constraint by
+                ignoring the phase angle and taking the absolute value of all load coefficients. Default False.
+
+        Returns:
+            List[complex]: Aggregate currents subject to the given constraints.
+        """
+        # Convert list of constraint id's to list of indices in constraint matrix
+        if constraints:
+            constraint_indices = [self.constraint_index[constraint_id] for constraint_id in constraints]
+        else:
+            constraint_indices = list(self.constraint_index.values())
+        if t:
+            schedule_length = len(t)
+            schedule_matrix = np.array([[load_currents[evse_id][i] for i in t] if evse_id in load_currents else [0] * schedule_length for evse_id, _ in sorted(self._EVSEs.items())])
+        else:
+            schedule_length = len(next(iter(load_currents.values())))
+            schedule_matrix = np.array([load_currents[evse_id] if evse_id in load_currents else [0] * schedule_length for evse_id, _ in sorted(self._EVSEs.items())])
+        if linear:
+            return complex(np.abs(self.constraint_matrix[constraint_indices]@schedule_matrix))
+        else:
+            # build vector of phase angles on EVSE
+            angle_coeffs = np.exp(1j*np.deg2rad(self.angles_vector))
+
+            # multiply schedule by angles matrix element-wise
+            shifted_schedule = (schedule_matrix.T * angle_coeffs).T
+
+            # multiply constraint matrix by current schedule, shifted by the phases
+            return self.constraint_matrix[constraint_indices]@shifted_schedule
+
+    def is_feasible(self, load_currents, t=0, linear=False):
+        """ Return if a set of current magnitudes for each load are feasible.
+
+        Args:
+            load_currents (Dict[str, List[number]]): Dictionary mapping load_ids to schedules of charging rates.
+            t (int): Index into the charging rate schedule where feasibility should be checked.
+            linear (bool): If True, linearize all constraints to a more conservative but easier to compute constraint by
+                ignoring the phase angle and taking the absolute value of all load coefficients. Default False.
+
+        Returns:
+            bool: If load_currents is feasible at time t according to this set of constraints.
+        """
+        # build schedule matrix, ensuring rows in order of EVSE list
+        schedule_lengths = set(len(x) for x in load_currents.values())
+        schedule_length = schedule_lengths.pop()
+        schedule_matrix = np.array([load_currents[evse_id] if evse_id in load_currents else [0] * schedule_length for evse_id, _ in sorted(self._EVSEs.items())])
```

Comment:
I think is_feasible should use constraint_currents to avoid code duplication unless we have a good reason not to?